packages/core/core/src/BundleGraph.js

```diff
@@ -874,42 +943,79 @@ export default class BundleGraph {
   }
 
   getBundlesInBundleGroup(bundleGroup: BundleGroup): Array<Bundle> {
-    return bundleGroup.bundleIds
-      .map(id => {
-        let b = nullthrows(this._graph.getNode(id));
-        invariant(b.type === 'bundle');
-        return b.value;
-      })
-      .reverse();
+    let bundles: Set<Bundle> = new Set();
+    let bundleGroupNode = nullthrows(
+      this._graph.getNode(getBundleGroupId(bundleGroup)),
+    );
+    for (let bundleNode of this._graph.getNodesConnectedFrom(
+      bundleGroupNode,
+      'bundle',
+    )) {
+      invariant(bundleNode.type === 'bundle');
+      bundles.add(bundleNode.value);
+
+      for (let referencedBundle of this.getReferencedBundles(
+        bundleNode.value,
+      )) {
+        bundles.add(referencedBundle);
+      }
+    }
+
+    return [...bundles];
   }
 
-  getSiblingBundles(bundle: Bundle): Array<Bundle> {
-    let siblings = new Set();
+  getReferencedBundles(
+    bundle: Bundle,
+    opts?: {|recursive: boolean|},
+  ): Array<Bundle> {
+    let recursive = opts?.recursive ?? true;
+    let referencedBundles = new Set();
+    let stack = [bundle];
+
+    let _bundle;
+    while ((_bundle = stack.pop())) {
```

Comment:
this custom dfs can probably be replaced by `graph.traverse` or `graph.dfs` along reference edges.